lang/cs/Org.Apache.REEF.IMRU/OnREEF/IMRUTasks/TaskCloseCoordinator.cs

```diff
@@ -0,0 +1,113 @@
+﻿// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+// 
+//   http://www.apache.org/licenses/LICENSE-2.0
+// 
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+using System;
+using System.Text;
+using System.Threading;
+using Org.Apache.REEF.Common.Tasks.Events;
+using Org.Apache.REEF.IMRU.OnREEF.Driver;
+using Org.Apache.REEF.IMRU.OnREEF.Parameters;
+using Org.Apache.REEF.Tang.Annotations;
+using Org.Apache.REEF.Utilities.Logging;
+
+namespace Org.Apache.REEF.IMRU.OnREEF.IMRUTasks
+{
+    /// <summary>
+    /// This class provides a method to handle Task close event. It is called from TaskCloseEventHandler. 
+    /// It also wraps flags to represent if the task should be closed and if the task has been stopped
+    /// so that to provide a coordination between the task and the close handler.  
+    /// </summary>
+    internal sealed class TaskCloseCoordinator
+    {
+        private static readonly Logger Logger = Logger.GetLogger(typeof(TaskCloseCoordinator));
+
+        /// <summary>
+        /// When a close event is received, this variable is set to 1. At the beginning of each task iteration,
+        /// if this variable is set to 1, the task will break from the loop and return from the Call() method.
+        /// </summary>
+        private long _shouldCloseTask = 0;
+
+        /// <summary>
+        /// Before the task is returned, this variable is set to 1.
+        /// Close handler will check this variable to decide if it needs to throw an exception.
+        /// </summary>
+        private long _isTaskStopped = 0;
```

Comment:
It is needed.
In HandleEvent method, waitToCloseEvent.Wait can come out either by "set" before the task returns properly, or time out while task is hung. In the end, we need to use this flag to decide if we should throw exception. If the task was returned properly, then no need to throw exception, the driver will receive CompletedTask event. If _isTaskStopped  is not 1, meaning the task was not returned, but hung or something is wrong, we need to throw exception for driver to get FailedTask event.